include/boost/beast/core/buffers_generator.hpp

```diff
@@ -0,0 +1,200 @@
+    to the stream's `write_some` function.
+
+    @param stream The stream to which the data is to be written.
+    The type must support the <em>SyncWriteStream</em> concept.
+
+    @param generator The generator to use.
+
+    @return The number of bytes written to the stream.
+
+    @throws system_error Thrown on failure.
+
+    @see BuffersGenerator
+*/
+template<
+    class SyncWriteStream,
+    class BuffersGenerator
+#if ! BOOST_BEAST_DOXYGEN
+    , typename std::enable_if<is_buffers_generator<
+        typename std::decay<BuffersGenerator>::
+            type>::value>::type* = nullptr
+#endif
+    >
+std::size_t
+write(
+    SyncWriteStream& stream,
+    BuffersGenerator&& generator);
+
+/** Write all output from a BuffersGenerator asynchronously to a
+    stream.
+
+    This function is used to write all of the buffers generated
+    by a caller-provided @ref BuffersGenerator to a stream. The
+    function call always returns immediately. The asynchronous
+    operation will continue until one of the following
+    conditions is true:
+
+    @li A call to the generator's `is_done` returns `false`.
+
+    @li An error occurs.
+
+    This operation is implemented in terms of zero or more calls
+    to the stream's `async_write_some` function, and is known as
+    a <em>composed operation</em>.  The program must ensure that
+    the stream performs no other writes until this operation
+    completes.
+
+    @param stream The stream to which the data is to be written.
+    The type must support the <em>SyncWriteStream</em> concept.
+
+    @param generator The generator to use.
+
+    @param handler The completion handler to invoke when the
+    operation completes. The implementation takes ownership of
+    the handler by performing a decay-copy. The equivalent
+    function signature of the handler must be:
+    @code
+    void handler(
+        error_code const& error,        // result of operation
+        std::size_t bytes_transferred   // the number of bytes written to the stream
+    );
+    @endcode
+    Regardless of whether the asynchronous operation completes
+    immediately or not, the handler will not be invoked from
+    within this function. Invocation of the handler will be
+    performed in a manner equivalent to using `net::post`.
+
+    @see BuffersGenerator
+*/
+template<
+    class AsyncWriteStream,
+    class BuffersGenerator,
+    class CompletionToken
+#if !BOOST_BEAST_DOXYGEN
+    , typename std::enable_if<is_buffers_generator<
+        BuffersGenerator>::value>::type* = nullptr
+#endif
+    >
+auto
+async_write(
+    AsyncWriteStream& stream,
+    BuffersGenerator generator,
+    CompletionToken&& token) ->
+        typename net::async_result<
+            typename std::decay<
+                CompletionToken>::type,
+            void(error_code, std::size_t)>::return_type;
+
+} // beast
+} // boost
+
+#include <boost/beast/core/impl/buffers_generator.hpp>
```

Comment:
You're a beast. How'd you spot that? (Corollary, how'd that compile?)

It was a new pitfall for me since to date I never worked in "boost super-project" layout. It won't happen again.

Reply:
I saw the word "include." It compiles because... I guess the build scripts throw enough include directories at the compiler to make it work.

Reply:
Reviewing other people's code (and also, reviewing my own code) by looking at a file from top to bottom helps sharpen your ability to spot things. I do review my own code from time to time I open a file and look through it, usually before I merge something but also sometimes randomly.

Reply:
I do that all the time with my own code. Apparently there's still the blind spot effect.